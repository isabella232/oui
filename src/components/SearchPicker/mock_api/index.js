import MOCK_DATA from './mock_data';

export const getSearchFunction = (DELAY = 500) => ({ query, per_page }) => {
  return new Promise(r => setTimeout(r, DELAY)).then(() => {
    return MOCK_DATA.filter(item => {
      return item.name.includes(query) || item.description.includes(query);
    }).slice(0, per_page);
  });
};
