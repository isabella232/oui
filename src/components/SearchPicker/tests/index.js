import React from 'react';
import { mount } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';

import { getSearchFunction } from '../mock_api';
import SearchPicker from '../';

const MOCK_API_DELAY = 10;

const searchFunction = getSearchFunction(MOCK_API_DELAY);

const getDelayedPromise = (delay = 0) => new Promise(r => setTimeout(r, delay));


describe('components/SearchPicker', () => {
  let component;
  let renderedData;
  let onItemSelectedStub;

  describe('default case', () => {
    beforeEach(() => {
      onItemSelectedStub = jest.fn();
      component = mount(
        <SearchPicker
          onItemSelected={ onItemSelectedStub }
          searchFunction={ searchFunction }
          supportedTypes={ ['feature'] }>
          {(data) => {
            renderedData = data;
            return (
              <div>
                {data.renderInput()}
                <p>A list of entities</p>
              </div>
            );
          }}
        </SearchPicker>
      );
    });

    it('should render component same as snapshot', () => {
      expect(mountToJson(component)).toMatchSnapshot();
    });

    it('should pass a default list of items on creation', () => {
      return getDelayedPromise().then(() => {
        component.find('input').simulate('click');
      })
        .then(() => getDelayedPromise(MOCK_API_DELAY))
        .then(() => {
          component.update();
          expect(renderedData.availableEntities).toHaveLength(5);
        });
    });

    it('should update the list when a search term is entered', () => {
      return getDelayedPromise().then(() => {
        component.find('input').simulate('mouseenter');
        component.find('input').simulate('click');
        component.find('input').simulate('input', { target: { value: '123' }});
      })
        // Account for our mock API delay and the debounce delay
        .then(() => getDelayedPromise(MOCK_API_DELAY + 10))
        .then(() => {
          component.update();
          expect(renderedData.availableEntities).toHaveLength(1);
          expect(renderedData.availableEntities[0].name).toContain('My feature 123');
        });
    });

    it('should pass isLoading and the searchQuery correctly', () => {
      expect(renderedData.isLoading).toBe(true);
      return getDelayedPromise(MOCK_API_DELAY).then(() => {
        expect(renderedData.searchQuery).toBe('');
        expect(renderedData.isLoading).toBe(false);
        component.find('input').simulate('mouseenter');
        component.find('input').simulate('click');
        component.find('input').simulate('input', { target: { value: '123' }});
      })
        // Wait for debounce
        .then(() => getDelayedPromise(10))
        .then(() => {
          expect(renderedData.searchQuery).toBe('123');
          expect(renderedData.isLoading).toBe(true);
        })
        // Account for our mock API delay and the debounce delay
        .then(() => getDelayedPromise(MOCK_API_DELAY + 10))
        .then(() => {
          component.update();
          expect(renderedData.isLoading).toBe(false);
        });
    });
  });
});
