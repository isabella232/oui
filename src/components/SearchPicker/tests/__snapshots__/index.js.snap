// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SearchPicker default case should render component same as snapshot 1`] = `
<withkeyboardTracker(undefined)
  additionalItems={0}
  onItemSelected={[MockFunction]}
  searchFunction={[Function]}
  supportedTypes={
    Array [
      "feature",
    ]
  }
>
  <SearchPicker
    currentFauxFocusIndex={0}
    handleKeyDown={[Function]}
    onItemSelected={[MockFunction]}
    searchFunction={[Function]}
    searchOptions={Object {}}
    selectedEntityIds={Array []}
    setItemCount={[Function]}
    setOnItemSelect={[Function]}
    supportedTypes={
      Array [
        "feature",
      ]
    }
  >
    <div>
      <[object Object]
        isFilter={true}
        isRequired={false}
        note={null}
        onInput={[Function]}
        onKeyDown={[Function]}
        placeholder="Search for features"
        testSection="filter-features-input"
        type="search"
        value=""
      >
        <input
          className="oui-text-input oui-text-input--search"
          data-oui-component={true}
          data-test-section="filter-features-input"
          onInput={[Function]}
          onKeyDown={[Function]}
          placeholder="Search for features"
          required={false}
          type="search"
          value=""
        />
      </[object Object]>
      <p>
        A list of entities
      </p>
    </div>
  </SearchPicker>
</withkeyboardTracker(undefined)>
`;
